controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -186,29 +208,37 @@ private void triggerReadyJobs(LockedApplication application) {
      */
     private boolean changesAvailable(Application application, JobStatus previous, JobStatus next) {
         if ( ! application.deploying().isPresent()) return false;
-        Change change = application.deploying().get();
 
-        if ( ! previous.lastSuccess().isPresent()) return false;
+        Change change = application.deploying().get();
 
-        if (change instanceof Change.VersionChange) {
+        if (change instanceof Change.VersionChange) { // Propagate upgrade while making sure we never downgrade
             Version targetVersion = ((Change.VersionChange)change).version();
-            if ( ! (targetVersion.equals(previous.lastSuccess().get().version())) )
-                return false; // version is outdated
-            // The below is checked again in allowedTriggering, right before actual triggering.
-            if (next != null && isOnNewerVersionInProductionThan(targetVersion, application, next.type()))
-                return false; // Don't downgrade
-        }
 
-        if (next == null) return true;
-        if ( ! next.lastSuccess().isPresent()) return true;
+            // Is the target version tested?
+            if ( ! lastSuccessfulIs(targetVersion, JobType.stagingTest, application))
+                return false;
+
+            // Is it this job's turn to upgrade?
+            if ( previous.type().isProduction() && ! isOnAtLeastProductionVersion(targetVersion, application, previous.type()))
+                return false;
```

Comment:
Hmm...

1. Production job `eu-1` deploys version _A_, but fails the verification (aborts or fails for the second time).
2. We enter this method again with `previous` set to `eu-1`, and then pass this `if` clause. 
3. The below `is`s don't stop us either, and we end up triggering `eu-2`, even though `eu-1` failed verification. 

I can't find anything to stop this from happening?

Reply:
I see this is a feature in the test scenario below, necessary to avoid deadlocks. Perhaps it would be better to differentiate between `isOnNewerVersion` and `isOnAtLeastVersion`, though?